src/metabase/driver/sql_jdbc/actions.clj

```diff
@@ -1,235 +1,277 @@
-                                {::incorrect-number-updated true
-                                 :number-updated           rows-updated
-                                 :query                    query
-                                 :sql                      (hformat/format update-hsql)
-                                 :status-code              400}))))))
-      (catch Exception e
-        (let [e-data (if (::incorrect-number-updated (ex-data e))
-                       (ex-data e)
-                       (parse-error driver conn e))]
-          (throw
-           (ex-info (or (ex-message e) "Update action error.")
-                    (assoc e-data :status-code 400))))))))
+                                :set (cast-values driver update-row (get-in query [:query :source-table]))))
+        sql-args     (sql.qp/format-honeysql driver update-hsql)]
+    (with-jdbc-transaction [conn database-id]
+      (try
+        ;; TODO -- this should probably be using [[metabase.driver/execute-write-query!]]
+        (let [rows-updated (first (jdbc/execute! {:connection conn} sql-args {:transaction? false}))]
+          (when-not (= rows-updated 1)
+            (throw (ex-info (tru "Sorry, this would update {0} rows, but you can only act on 1" rows-updated)
+                            {::incorrect-number-updated true
+                             :number-updated            rows-updated
+                             :query                     query
+                             :sql                       (sql.qp/format-honeysql driver update-hsql)
+                             :status-code               400})))
+          {:rows-updated [1]})
+        (catch Exception e
+          (let [e-data (if (::incorrect-number-updated (ex-data e))
+                         (ex-data e)
+                         (parse-error driver conn e))]
+            (throw
+             (ex-info (or (ex-message e) "Update action error.")
+                      (assoc e-data :status-code 400)))))))))
+
+(defn- select-created-row
+  "H2 and MySQL are dumb and `RETURN_GENERATED_KEYS` only returns the ID of the newly created row. This function will
+  `SELECT` the newly created row."
+  [driver raw-hsql conn result]
+  (let [select-hsql     (assoc raw-hsql
+                               :select [:*]
+                               ;; :and with a single clause will be optimized in HoneySQL
+                               :where (into [:and]
+                                            (for [[col val] result]
+                                              [:= (keyword col) val])))
+        select-sql-args (sql.qp/format-honeysql driver select-hsql)]
+    (log/tracef ":row/create SELECT HoneySQL:\n\n%s" (u/pprint-to-str select-hsql))
+    (log/tracef ":row/create SELECT SQL + args:\n\n%s" (u/pprint-to-str select-sql-args))
+    (first (jdbc/query {:connection conn} select-sql-args {:identifiers identity, :transaction? false}))))
```

Comment:
I refactored the `:row/create` method a bit because it was kinda big (35 lines) and split some of it out into separate functions